src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -75,13 +75,14 @@ function frule(
     return F, ∂F
 end
 
-function _lu_pullback(ΔF::Tangent, A, pivot, F)
+# these functions are defined outside the rrule because otherwise type inference breaks
+# see https://github.com/JuliaLang/julia/issues/40990
+function _lu_pullback(ΔF::Tangent, m, n, eltypeA, pivot, F)
     Δfactors = ΔF.factors
     Δfactors isa AbstractZero && return (NoTangent(), Δfactors, NoTangent())
     factors = F.factors
-    ∂factors = eltype(A) <: Real ? real(Δfactors) : Δfactors
+    ∂factors = eltypeA <: Real ? real(Δfactors) : Δfactors
```

Comment:
Check that this works right.
I suspect you need to do

`_lu_pullback(ΔF::Tangent, m, n, :Type{eltypeA}, pivot, F) where eltypeA` 
or julia will not specialize on this and it will result in it being type-unstable.

Reply:
It does seem to work right
```julia
julia> y, pb = rrule(lu, randn(T, m, n), pivot)

julia> @code_warntype pb(ChainRulesTestUtils.rand_tangent(y))
Variables
  #self#::ChainRules.var"#lu_pullback#1571"{Matrix{Float64}, Val{true}, LU{Float64, Matrix{Float64}}, Int64, Int64}
  ȳ::Tangent{LU{Float64, Matrix{Float64}}, NamedTuple{(:factors, :ipiv, :info), Tuple{Matrix{Float64}, Vector{NoTangent}, NoTangent}}}

Body::Tuple{NoTangent, Matrix{Float64}, NoTangent}
1 ─ %1 = Core.getfield(#self#, :m)::Int64
│   %2 = Core.getfield(#self#, :n)::Int64
│   %3 = Core.getfield(#self#, :A)::Matrix{Float64}
│   %4 = ChainRules.eltype(%3)::Core.Const(Float64)
│   %5 = Core.getfield(#self#, :pivot)::Core.Const(Val{true}())
│   %6 = Core.getfield(#self#, :F)::LU{Float64, Matrix{Float64}}
│   %7 = ChainRules._lu_pullback(ȳ, %1, %2, %4, %5, %6)::Core.PartialStruct(Tuple{NoTangent, Matrix{Float64}, NoTangent}, Any[Core.Const(NoTangent()), Matrix{Float64}, Core.Const(NoTangent())])
└──      return %7
```

but also the `@inferred` tests should have caught a regression, right?

Reply:
nice, good.
Yes they should ahve